esp/src/src-react/components/PackageMapParts.tsx

```diff
@@ -0,0 +1,158 @@
+            }
+         },
+    });
+
+    const refreshTable = React.useCallback((clearSelection = false) => {
+        grid?.set("query", gridQuery);
+        if (clearSelection) {
+            grid?.clearSelection();
+        }
+    }, [grid, gridQuery]);
+
+    //  Command Bar ---
+    const buttons = React.useMemo((): ICommandBarItemProps[] => [
+        {
+            key: "refresh", text: nlsHPCC.Refresh, iconProps: { iconName: "Refresh" },
+            onClick: () => refreshTable()
+        },
+        { key: "divider_1", itemType: ContextualMenuItemType.Divider, onRender: () => <ShortVerticalDivider /> },
+        {
+            key: "add", text: nlsHPCC.Add,
+            onClick: () => setShowAddPartForm(true)
+        },
+        {
+            key: "delete", text: nlsHPCC.RemovePart, disabled: !uiState.hasSelection,
+            onClick: () => {
+                if (confirm(nlsHPCC.YouAreAboutToDeleteThisPart)) {
+                    selection.forEach((item, idx) => {
+                        WsPackageMaps.RemovePartFromPackageMap({
+                            request: {
+                                PackageMap: name.split("::")[1],
+                                Target: _package.Target,
+                                PartName: item.Part
+                            }
+                        }).then(({ RemovePartFromPackageMapResponse, Exceptions }) => {
+                            if (RemovePartFromPackageMapResponse?.status?.Code === 0) {
+                                gridStore.remove(item.Part);
+                                refreshTable();
+                            } else if (Exceptions?.Exception.length > 0) {
+                                setShowError(true);
+                                setErrorMessage(Exceptions?.Exception[0].Message);
+                            }
+                        });
+                    });
+                }
+            }
+        },
+        { key: "divider_2", itemType: ContextualMenuItemType.Divider, onRender: () => <ShortVerticalDivider /> },
+        {
+            key: "getPart", text: nlsHPCC.GetPart, disabled: !uiState.hasSelection,
+            onClick: () => {
+                if (selection.length === 1) {
+                    pushUrl(`/packagemaps/${name}/parts/${selection[0].Part}`);
+                } else {
+                    selection.forEach(item => {
+                        window.open(`#/packagemaps/${name}/parts/${item.Part}`, "_blank");
+                    });
+                }
+            }
+        },
+    ], [_package, gridStore, name, refreshTable, selection, uiState.hasSelection]);
+
+    React.useEffect(() => {
+        WsPackageMaps.getPackageMapById({ packageMap: name }).then(({ GetPackageMapByIdResponse }) => {
+            const xml = parser.parse(GetPackageMapByIdResponse?.Info);
+            const parts = [...xml.getElementsByTagName("Part")].map(part => {
```

Comment:
Not sure the outer `[... ]` adds anything given its about to be mapped anyway?

Reply:
Needed to be converted before calling map(), getElementsByTagName returns an HTMLCollection which is array-like, but doesn't have most of the functions of Array.prototype